aten/src/ATen/core/jit_type.h

```diff
@@ -324,25 +324,60 @@ inline c10::optional<T> merge_primitive(
   return c10::optional<T>{};
 }
 
-// `VaryingShape` tracks if individual dimensions or a rank vary across
-// profiled runs. A *varying* or *dynamic* dimension is expressed as
-// an empty c10::optional in `sizes_`. If a rank is dynamic, the entire
-// `sizes_` becomes the empty optional.
+struct CAFFE2_API ShapeSymbol {
+    
+    
+    // needed for use in `std::map`
+    ShapeSymbol(): value_(-1) {}
+    bool is_static() const { return value_ >= 0; };  
+    bool operator ==(const ShapeSymbol &b) const { return value_ == b.value_; }
+    bool operator <(const ShapeSymbol &b) const { return value_ < b.value_; }
+    // this will be removed in the next PR since we will also remove `isCompatibleWithInCurrentExecutionContext`
+    // as TensorType::merge will be different with dynamic shapes.
+    explicit operator int() const { return value_; }
+
+    static ShapeSymbol fromStaticSize(int64_t val) {return ShapeSymbol(val); }
+    int64_t static_size() const {
+      TORCH_CHECK(is_static());
+      return value_;
+    };
+
+    static ShapeSymbol newSymbol () {return ++num_symbols; };
```

Comment:
Fresh symbols should have negative values, right? Otherwise the symbol gets a fixed size?